Parse/ParseClientConfiguration.m

```diff
@@ -0,0 +1,125 @@
+#import "PFApplication.h"
+#import "PFCommandRunningConstants.h"
+#import "PFFileManager.h"
+#import "PFHash.h"
+
+@implementation ParseClientConfiguration
+
+///--------------------------------------
+#pragma mark - Init
+///--------------------------------------
+
++ (instancetype)emptyConfiguration {
+    return [super new];
+}
+
+- (instancetype)initWithBlock:(void (^)(id<ParseMutableClientConfiguration>))configurationBlock {
+    self = [super init];
+    if (!self) return nil;
+
+    _networkRetryAttempts = PFCommandRunningDefaultMaxAttemptsCount;
+
+    configurationBlock(self);
+
+    return self;
+}
+
++ (instancetype)configurationWithBlock:(void (^)(id<ParseMutableClientConfiguration>))configurationBlock {
+    return [[self alloc] initWithBlock:configurationBlock];
+}
+
+///--------------------------------------
+#pragma mark - Properties
+///--------------------------------------
+
+- (void)setApplicationId:(NSString *)applicationId {
+    PFConsistencyAssert([applicationId length], @"'applicationId' should not be nil.");
+    _applicationId = [applicationId copy];
+}
+
+- (void)setClientKey:(NSString *)clientKey {
+    PFConsistencyAssert([clientKey length], @"'clientKey' should not be nil.");
+    _clientKey = [clientKey copy];
+}
+
+- (void)setApplicationGroupIdentifier:(NSString *)applicationGroupIdentifier {
+    PFConsistencyAssert(
+        applicationGroupIdentifier == nil ||
+        [PFFileManager isApplicationGroupContainerReachableForGroupIdentifier:applicationGroupIdentifier],
+        @"ApplicationGroupContainer is unreachable. Please double check your Xcode project settings."
+    );
+
+    _applicationGroupIdentifier = [applicationGroupIdentifier copy];
+}
+
+- (void)setContainingApplicationBundleIdentifier:(NSString *)containingApplicationBundleIdentifier {
+    PFConsistencyAssert(containingApplicationBundleIdentifier == nil ||
+                        [PFApplication currentApplication].extensionEnvironment,
+                        @"'containingApplicationBundleIdentifier' must be non-nil in extension environment");
+
+    PFConsistencyAssert(containingApplicationBundleIdentifier != nil ||
+                        ![PFApplication currentApplication].extensionEnvironment,
+                        @"'containingApplicationBundleIdentifier' must be nil in non-extension environment");
+
+    _containingApplicationBundleIdentifier = containingApplicationBundleIdentifier;
+}
+
+- (void)_resetDataSharingIdentifiers {
+    _applicationGroupIdentifier = nil;
+    _containingApplicationBundleIdentifier = nil;
+}
+
+///--------------------------------------
+#pragma mark - NSObject
+///--------------------------------------
+
+- (NSUInteger)hash {
+    return PFIntegerPairHash(self.applicationId.hash, self.clientKey.hash);
+}
+
+- (BOOL)isEqual:(id)object {
+    if (![object isKindOfClass:[ParseClientConfiguration class]]) {
+        return NO;
+    }
+
+#define EQUAL_OBJECTS(a, b) ((a) == (b) || [(a) isEqual:(b)])
```

Comment:
Can we avoid doing this? Inline defines are quite sad.
There is `PFObjectUtilities.+isObject:equalToObject:` for this very reason :grin: